### PR TITLE
Implement `hash_map` macro 

### PR DESCRIPTION
**Pull request by stifskere**
Implementation of https://github.com/rust-lang/rust/issues/144032

Implements the `hash_map` macro under `std/src/macros.rs`.

### CODE REVIEW

**Comment by NyxCode** [comment]
Wouldn't using one of `HashMap::with_capacity()`, `HashMap::reserve()`, `FromIterator<(K, V)>`, `Extend<(K, V)>` or `From<[(K, V); N]>` be better, avoiding unnecessary allocations?

**Comment by stifskere** [comment]
> Wouldn't using one of `HashMap::with_capacity()`, `HashMap::reserve()`, `FromIterator<(K, V)>`, `Extend<(K, V)>` or `From<[(K, V); N]>` be better, avoiding unnecessary allocations?

I was thinking about `HashMap::with_capacity()` but I couldn't see how I could convert a repetition into a number.

`From...` would essentially be the same, just reallocation, so for clarity I just expanded into `insert`.

~I could see that `0 $(+ 1)*` could be evaluated constantly by the compiler, thus we could get a size, but I'm not sure if it's the best way.~
It could not be the best way, because that expansion doesn't at all include a reference to the repetition declaration, so it can't be. There is no logical way of getting the size with a macro_rules unless with more complexity we make a recursive macro, *which would limit the hash_map size to 254* without extra compiler parameters.

~I will add a question about the size in the tracking issue.~

**Comment by NyxCode** [comment]
All the options i mentioned should be equivalent to `with_capacity` as far as I can tell, though I don't know if the optimizer can look through all options equally well (e.g. wrt. stack usage).  
Both these should work, for example:
```rust
let len = 1 $(+ {stringify!($k); 1})*;
let mut map = HashMap::with_capacity(len);
$( map.insert($k, $v); )*
```
```rust
HashMap::from([$(($k, $v)),*])
```

File: library/std/src/macros.rs [hunk 1]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
...
+///
+/// Note that since this macro only generates [`HashMap::insert`]
+/// calls, the move semantics for the values are the same as
+/// the insert method.
+///
+/// [`HashMap`]: crate::collections::HashMap
+/// [`HashMap::new`]: crate::collections::HashMap::new
+/// [`HashMap::insert`]: crate::collections::HashMap::insert
+#[macro_export]
+#[allow_internal_unstable(hint_must_use)]
```
Comment (Noratrieb) [inline library/std/src/macros.rs]: you don't use of this feature (maybe it would make sense to?)

File: library/std/src/macros.rs [hunk 2]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
...
+#[unstable(feature = "hash_map_macro", issue = "144032")]
+macro_rules! hash_map {
+    () => {{
+        ::std::collections::HashMap::new()
+    }};
+
+    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
+        let mut __map = ::std::collections::HashMap::new();
```
Comment (Noratrieb) [inline library/std/src/macros.rs]: you can just call it `map`, no underscores needed thanks to hygiene

File: library/std/src/macros.rs (hunk 1 above)
Comment (stifskere) [reply library/std/src/macros.rs]: I added it because it makes sense that if you create a hash map you must use it, otherwise don't create it, I don't see the point otherwise.

**Comment by stifskere** [comment]
> All the options i mentioned should be equivalent to `with_capacity` as far as I can tell, though I don't know if the optimizer can look through all options equally well (e.g. wrt. stack usage). Both these should work, for example:
> 
> ```rust
> let len = 1 $(+ {stringify!($k); 1})*;
> let mut map = HashMap::with_capacity(len);
> $( map.insert($k, $v); )*
> ```
> 
> ```rust
> HashMap::from([$(($k, $v)),*])
> ```

I think I'd rather go for intrinsics, except since `HashMap` is a wrapper, I don't think there are many. I feel that if someone decides to expand the macro via IDE or cargo expand, they will find a lot of strings and be like "???". Probably open an issue later.

Otherwise, we could generate a constant count probably, the compiler should optimize this, and it should not have runtime cost, but a bit extra of constant cost, which may mean that if there are a lot of invocations the compiler might go a bit slower.

```rust
macro_rules! hash_map {
    () => {{
        ::std::collections::HashMap::new()
    }};

    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
        let mut map = ::std::collections::HashMap::with_capacity(
            const { [$({ let _ = &$key; () }),*].len() }
        );

        $( map.insert($key, $value); )*
        map
    }}
}
```

image

**Comment by Noratrieb** [comment]
I don't think intrinsics make sense here, this is a library type. I think the repeat insert method makes the most sense as it doesn't have the potential stack usage problem of from. Repeat insert is also how extend is implemented under the hood: https://github.com/rust-lang/hashbrown/blob/670213fb32d208759d0331996894e05604de0c18/src/map.rs#L4562.

with_capacity definitely makes sense though. You can count the elements like this: https://lukaswirth.dev/tlborm/decl-macros/building-blocks/counting.html

**Comment by Noratrieb** [comment]
Oh actually you might be able to use the unstable macro_metavar_expr_count for the counting too, I forgot about that :D

**Comment by stifskere** [comment]
> I don't think intrinsics make sense here, this is a library type. I think the repeat insert method makes the most sense as it doesn't have the potential stack usage problem of from. Repeat insert is also how extend is implemented under the hood: https://github.com/rust-lang/hashbrown/blob/670213fb32d208759d0331996894e05604de0c18/src/map.rs#L4562.
> 
> with_capacity definitely makes sense though. You can count the elements like this: https://lukaswirth.dev/tlborm/decl-macros/building-blocks/counting.html

I would go for the [slice length](https://lukaswirth.dev/tlborm/decl-macros/building-blocks/counting.html#slice-length) approach, but resolved constantly, which is technically what I'm doing just that instead of making a variable using `replace_expr!`

So probably
```rust
macro_rules! hash_map {
    () => {{
        ::std::collections::HashMap::new()
    }};

    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
        let mut map = ::std::collections::HashMap::with_capacity(
            const { [$($crate::hash_map!(@internal count $key)),*].len() }
        );

        $( map.insert($key, $value); )*
        map
    }};

    (@internal count $i:tt) => {()}
}
```
This would work best if there is some sort of thing inside the standard library so that the last branch is only used internally in an explicit way, which would also filter out the usage inside the second branch.

**Comment by Noratrieb** [comment]
> Oh actually you might be able to use the unstable macro_metavar_expr_count for the counting too, I forgot about that :D

You should try this, which I only remembered after posting the first message

**Comment by stifskere** [comment]
> > Oh actually you might be able to use the unstable macro_metavar_expr_count for the counting too, I forgot about that :D
> 
> You should try this, which I only remembered after posting the first message 

The problem with this is that it would delay stabilization a lot, so I'm not sure, I don't think adding unstable to unstable helps

**Comment by Noratrieb** [comment]
You're allowed to use unstable features internally with `#[allow_internal_unstable()]` and listing out the feature names there, on the macro.

**Comment by tmke8** [comment]
Would it be possible to use `HashMap::default()` instead of `HashMap::new()` so that this macro can be used with non-default hashing algorithms as well? (Though this might cause type inference problems...)

EDIT: or rather

```rust
HashMap::with_capacity_and_hasher(capacity, Default::default())
```
for specifying capacity.

**Comment by stifskere** [comment]
> You're allowed to use unstable features internally with `#[allow_internal_unstable()]` and listing out the feature names there, on the macro.

Ok, I will check that now

**Comment by stifskere** [comment]
> Would it be possible to use `HashMap::default()` instead of `HashMap::new()` so that this macro can be used with non-default hashing algorithms as well? (Though this might cause type inference problems...)
> 
> EDIT: or rather
> 
> ```rust
> HashMap::with_capacity_and_hasher(capacity, Default::default())
> ```
> 
> for specifying capacity.

We are discussing to use `with_capacity` instead, If you could give your opinion on getting capacity from a macro repetition, possibly would be better, because knowing the size before creation avoids reallocation and makes the thing faster overall.

File: library/std/src/macros.rs (hunk 1 above)
Comment (tmke8) [reply library/std/src/macros.rs]: (Noratrieb's point was that you should actually use `must_use` within the macro. This line only makes it so that using `must_use` within the macro is *allowed*.)

**Comment by stifskere** [comment]
@Noratrieb Sorry for the inactivity, I was looking through if I could find the `macro_metavar_expr_count` feature, but it's not anywhere, nor there is documentation for it. I looked in issues, in the rust nightly book, in other pull requests... I couldn't find anything really.

**Comment by stifskere** [comment]
I personally don't find a constant evaluation that consuming, nor for the compiler, it's just allocating an array of sized 0 elements with 1 byte alignment and then counting it.

**Comment by stifskere** [comment]
> Would it be possible to use `HashMap::default()` instead of `HashMap::new()` so that this macro can be used with non-default hashing algorithms as well? (Though this might cause type inference problems...)
> 
> EDIT: or rather
> 
> ```rust
> HashMap::with_capacity_and_hasher(capacity, Default::default())
> ```
> 
> for specifying capacity.

I'm not sure if I could think of a syntax for that to be added, you could comment your proposal on the tracking issue.

**Comment by slanterns** [comment]
https://rust-lang.github.io/rfcs/3086-macro-metavar-expr.html
https://github.com/rust-lang/rust/tree/master/tests/ui/macros/metavar-expressions

**Comment by stifskere** [comment]
> https://rust-lang.github.io/rfcs/3086-macro-metavar-expr.html https://github.com/rust-lang/rust/tree/master/tests/ui/macros/metavar-expressions

Thanks for this, I was searching for `macro-metavar-expr-count` for which I didn't find anything, I tested in my computer and this should be working now.

**Comment by tmke8** [comment]
> I'm not sure if I could think of a syntax for that to be added, you could comment your proposal on the tracking issue.

I didn't mean to change the syntax, just the implementation:

```rust
macro_rules! hash_map {

    () => {{
        ::std::collections::HashMap::new()
    }};

    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
        let mut map = ::std::collections::HashMap::with_capacity_and_hasher(
            ${count($key)} , Default::default()
        );
        $( map.insert($key, $value); )*
        map
    }}
}

let map: FxHashMap = hash_map!["foo".to_string() => 0];
```

Though as I said I'm not sure about type inference. But you're right I should post in the tracking issue instead.

File: library/std/src/macros.rs [hunk 3]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
...
+/// a [`HashMap::insert`] call for each element.
+///
+/// This macro allows trailing commas.
+///
+/// This macro is used by concatenating comma separated sequences
+/// of `<key:expr> => <value:expr>`.
+///
+/// An example usage of this macro could be the following
+///
+/// ```rust
```
Comment (Noratrieb) [inline library/std/src/macros.rs]:
```suggestion
/// # Examples
///
/// ```
```

File: library/std/src/macros.rs [hunk 4]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
...
+/// #![feature(hash_map_macro)]
+///
+/// let map = hash_map!{
+///     "key" => "value",
+///     "key1" => "value1"
+/// };
+///
+/// for (key, value) in map {
+///     println!("{key:?} => {value:?}");
+/// }
```
Comment (Noratrieb) [inline library/std/src/macros.rs]: can you replace this with assertions instead? to make it clearer what the result is like (though it is pretty obvious) and also serve as a test.

File: library/std/src/macros.rs [hunk 5]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
...
+///
+/// This macro is used by concatenating comma separated sequences
+/// of `<key:expr> => <value:expr>`.
+///
+/// An example usage of this macro could be the following
+///
+/// ```rust
+/// #![feature(hash_map_macro)]
+///
+/// let map = hash_map!{
```
Comment (Noratrieb) [inline library/std/src/macros.rs]:
```suggestion
/// let map = hash_map! {
```
rustfmt puts a space before braced macro invocations.

File: library/std/src/macros.rs [hunk 6]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
         ($($crate::dbg!($val)),+,)
     };
 }
+
+/// Creates a [`HashMap`] containing the arguments.
+///
+/// This macro creates or either an empty [`HashMap`] using
+/// [`HashMap::new`] if there are no arguments, or generates
+/// a [`HashMap::insert`] call for each element.
```
Comment (Noratrieb) [inline library/std/src/macros.rs]: I'm not sure it's useful to be this precise about `insert`, I would probably avoid mentioning it. Mentioning it this clearly may imply that it's a guarantee (not sure how someone would rely on this though) which it is not. Anyone curious to see the implementation can view the source to see the `insert`.

File: library/std/src/macros.rs [hunk 7]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
...
+#[macro_export]
+#[allow_internal_unstable(macro_metavar_expr)]
+#[unstable(feature = "hash_map_macro", issue = "144032")]
+macro_rules! hash_map {
+    () => {{
+        ::std::collections::HashMap::new()
+    }};
+
+    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
+        let mut map = ::std::collections::HashMap::with_capacity( ${count($key)} );
```
Comment (Noratrieb) [inline library/std/src/macros.rs]:
```suggestion
        let mut map = $crate::collections::HashMap::with_capacity( ${count($key)} );
```

File: library/std/src/macros.rs [hunk 8]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
...
+///
+/// [`HashMap`]: crate::collections::HashMap
+/// [`HashMap::new`]: crate::collections::HashMap::new
+/// [`HashMap::insert`]: crate::collections::HashMap::insert
+#[macro_export]
+#[allow_internal_unstable(macro_metavar_expr)]
+#[unstable(feature = "hash_map_macro", issue = "144032")]
+macro_rules! hash_map {
+    () => {{
+        ::std::collections::HashMap::new()
```
Comment (Noratrieb) [inline library/std/src/macros.rs]:
```suggestion
        $crate::collections::HashMap::new()
```

File: library/std/src/macros.rs [hunk 9]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
...
+/// };
+///
+/// for (key, value) in map {
+///     println!("{key:?} => {value:?}");
+/// }
+/// ```
+///
+/// Note that since this macro only generates [`HashMap::insert`]
+/// calls, the move semantics for the values are the same as
+/// the insert method.
```
Comment (Noratrieb) [inline library/std/src/macros.rs]: I'm also not sure whether this sentence adds useful information.
Maybe you can add a sentence like this at the end of the top paragraph:

> The key and value are moved into the hashmap.

although I would consider the ownership semantics here fairly obvious, but a short sentence like the one above probably wouldn't hurt.

**Review by Noratrieb — COMMENTED** [review]
Looks like you figured out the `macro_metavar_expr`, that's great. It looks a lot better with `count` :). I see that it doesn't quite work though yet, I'm not sure why exactly.. it might be that the feature gating is not handled in the compiler in a way that makes `allow_internal_unstable` work..

I took a closer look at the implementation and documentation and wrote down a few minor notes. Additionally, can you add a few tests for example with trailing commas or an empty invocation?

Overall it looks good, so after you address the feedback and add the tests we can merge it (please do squash the commits into one before that though)!

File: library/std/src/macros.rs [hunk 10]
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
...
+///
+/// [`HashMap`]: crate::collections::HashMap
+/// [`HashMap::new`]: crate::collections::HashMap::new
+/// [`HashMap::insert`]: crate::collections::HashMap::insert
+#[macro_export]
+#[allow_internal_unstable(macro_metavar_expr)]
```
Comment (Noratrieb) [inline library/std/src/macros.rs]: I've tried to debug locally why the `allow_internal_unstable` doesn't work here and it seems like it's due to the time where the feature gate is checked during declarative macro compilation. I'll continue looking into it.

Comment (Noratrieb) [reply library/std/src/macros.rs]: Yeah it looks like I can't make easily make this work. I asked on Zulip if anyone else does: https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/allow_internal_unstable.28macro_metavar_expr.29

File: library/std/src/macros.rs (hunk 6 above)
Comment (stifskere) [reply library/std/src/macros.rs]: Should I mention the move semantics in some other way, or simply skip this completely?

File: library/std/src/macros.rs (hunk 10 above)
Comment (stifskere) [reply library/std/src/macros.rs]: > Yeah it looks like I can't make easily make this work. I asked on Zulip if anyone else does: https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/allow_internal_unstable.28macro_metavar_expr.29

For what you sent in [`#t-compiler > allow_internal_unstable(macro_metavar_expr) @ 💬`](https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/allow_internal_unstable.28macro_metavar_expr.29/near/532162756) looks like the compiler only cares about the compiled code and not about the macro itself for what I see

But a little bit above this macro in the same file I notice there is the `eprintln!` macro which has the `print_internals` and `format_args_nl` flags, suggesting that it may actually work, and the problem may be isolated in `macro_metavar_expr`.

Comment (stifskere) [reply library/std/src/macros.rs]: Right, forget the anterior comment, I misread, should I revert this to static evaluation or a simple `new`?

Comment (Noratrieb) [reply library/std/src/macros.rs]: Yeah, I think reverting away from using the feature is the best idea. In this case, switching to slice length counting probably makes the most sense? Please create a `#[doc(hidden)] #[unstable(feature = "hash_map_internals", issue = "none")]` helper macro instead of an `@internal` match arm though (and then use `#[allow_internal_unstable(hash_map_internals)]`.

Comment (stifskere) [reply library/std/src/macros.rs]: is it OK if I make this macro instead for general use? or it would require another ACP -> PR -> ...?

```rust
macro_rules! replace_expr {
    ($_t:tt $sub:expr) => {$sub};
}
```

Comment (Noratrieb) [reply library/std/src/macros.rs]: Adding new public macros would need another ACP and shouldn't be done here, but it seems unlikely that such a macro would be accepted. Just create an unstable doc(hidden) helper, which you can make a bit more generic if you want to (but I probably wouldn't do that)

**Comment by stifskere** [comment]
I will squash them when this is ready to merge, this way I don't need to do that multiple times

**Comment by stifskere** [comment]
I yesterday was running solely with a coffee lol.

**Comment by stifskere** [comment]
I was yesterday solely running on a monster energy drink, I now realize how bad that was.

**Comment by stifskere** [comment]
I think a macro is not a good idea, we semingly end with the same feature gate problem, I'l make some sort of hack with a const function instead, but without lsp I need to investigate where I can put that

**Comment by Noratrieb** [comment]
In this case you're just missing another `allow_internal_unstable` on your `repetition_utils`, since `allow_internal_unstable` isn't transitive:
```patch
diff --git a/library/std/src/macros.rs b/library/std/src/macros.rs
index 7055168e0a6..455d0ca9e24 100644
--- a/library/std/src/macros.rs
+++ b/library/std/src/macros.rs
@@ -383,6 +383,7 @@ macro_rules! dbg {
 #[doc(hidden)]
 #[macro_export]
 #[unstable(feature = "hash_map_internals", issue = "none")]
+#[allow_internal_unstable(hash_map_internals)]
 macro_rules! repetition_utils {
     (@count $($tokens:tt),*) => {{
         [$($crate::repetition_utils!(@replace $tokens => ())),*].len()
```
This makes it work for me.

**Comment by stifskere** [comment]
> In this case you're just missing another `allow_internal_unstable` on your `repetition_utils`, since `allow_internal_unstable` isn't transitive:
> 
> ```diff
> diff --git a/library/std/src/macros.rs b/library/std/src/macros.rs
> index 7055168e0a6..455d0ca9e24 100644
> --- a/library/std/src/macros.rs
> +++ b/library/std/src/macros.rs
> @@ -383,6 +383,7 @@ macro_rules! dbg {
>  #[doc(hidden)]
>  #[macro_export]
>  #[unstable(feature = "hash_map_internals", issue = "none")]
> +#[allow_internal_unstable(hash_map_internals)]
>  macro_rules! repetition_utils {
>      (@count $($tokens:tt),*) => {{
>          [$($crate::repetition_utils!(@replace $tokens => ())),*].len()
> ```
> 
> This makes it work for me.

Done, I added it, let's see if tests run correctly now.

By the way, what's the way you test whether tests fail and whether stability is correct locally? Because I would like to do the same instead of relying on running the workflows. I have read some about the `./x` bootstrap, I need to figure out how to configure it locally.

File: library/std/src/macros.rs (hunk 6 above)
Comment (Noratrieb) [reply library/std/src/macros.rs:399]: This is my attempt at the documentation, I rephrased a few things to attempt to make it a bit clearer and focus on the more important information:

```rust
/// Creates a [`HashMap`] containing the arguments.
/// 
/// `hash_map!` allows specifying the entries that make
/// up the [`HashMap`] where the key and value are separated by a `=>`.
/// 
/// The entries are separated by commas, where a trailing comma can be used.
/// 
/// It is semantically equivalent to using repeated [`HashMap::insert`]
/// on a newly created hashmap.
/// `hash_map!` will attempt to avoid repeated reallocations by using [`with_capacity`].
///
/// # Examples
///
/// ```rust
/// #![feature(hash_map_macro)]
///
/// let map = hash_map! {
///     "key" => "value",
///     "key1" => "value1"
/// };
///
/// assert_eq!(map.get("key"), Some(&"value"));
/// assert_eq!(map.get("key1"), Some(&"value1"));
/// ```
///
/// ```rust
/// #![feature(hash_map_macro)]
///
/// let map = hash_map! {
///     "key" => "value", // notice the ,
/// };
///
/// assert_eq!(map.get("key"), Some(&"value"));
/// assert_eq!(map.get("no-key"), None);
/// ```
///
/// [`HashMap`]: crate::collections::HashMap
/// [`HashMap::insert`]: crate::collections::HashMap::insert
/// [`with_capacity`]: crate::collections::HashMap::with_capacity
```

You can adapt some stuff from there (or take the whole thing if you think it's good). We don't necessarily have to make the documentation perfect in this PR, but we can try a bit to make it nice

**Comment by Noratrieb** [comment]
I can recommend reading this to see how to build and run the compiler with a standard library: https://rustc-dev-guide.rust-lang.org/building/quickstart.html
I did `./x build --stage 1 library` to build the compiler and have it linked up as mentioned in the guide to then be able to do `rustc +dev mytest.rs` to test building the file (where I invoke the macro).

File: library/std/src/macros.rs (hunk 6 above)
Comment (Noratrieb) [reply library/std/src/macros.rs]: This is my attempt at the documentation, I rephrased a few things to attempt to make it a bit clearer and focus on the more important information:

```rust
/// Creates a [`HashMap`] containing the arguments.
/// 
/// `hash_map!` allows specifying the entries that make
/// up the [`HashMap`] where the key and value are separated by a `=>`.
/// 
/// The entries are separated by commas, where a trailing comma can be used.
/// 
/// It is semantically equivalent to using repeated [`HashMap::insert`]
/// on a newly created hashmap.
/// `hash_map!` will attempt to avoid repeated reallocations by using [`with_capacity`].
///
/// # Examples
///
/// ```rust
/// #![feature(hash_map_macro)]
///
/// let map = hash_map! {
///     "key" => "value",
///     "key1" => "value1"
/// };
///
/// assert_eq!(map.get("key"), Some(&"value"));
/// assert_eq!(map.get("key1"), Some(&"value1"));
/// ```
///
/// ```rust
/// #![feature(hash_map_macro)]
///
/// let map = hash_map! {
///     "key" => "value", // notice the ,
/// };
///
/// assert_eq!(map.get("key"), Some(&"value"));
/// assert_eq!(map.get("no-key"), None);
/// ```
///
/// [`HashMap`]: crate::collections::HashMap
/// [`HashMap::insert`]: crate::collections::HashMap::insert
/// [`with_capacity`]: crate::collections::HashMap::with_capacity
```

You can adapt some stuff from there (or take the whole thing if you think it's good). We don't necessarily have to make the documentation perfect in this PR, but we can try a bit to make it nice

**Comment by Noratrieb** [comment]
(somehow github doesn't show it nicely despite me leaving it as a review comment but i added a comment here in case it got lost in the ui: https://github.com/rust-lang/rust/pull/144070#discussion_r2248866788)

**Comment by stifskere** [comment]
> (somehow github doesn't show it nicely despite me leaving it as a review comment but i added a comment here in case it got lost in the ui: [#144070 (comment)](https://github.com/rust-lang/rust/pull/144070#discussion_r2248866788))

I believe it didn't show up because it was solved, I didn't see the comment at all, If you could re-review the actual documentation, I changed it a bit.

**Comment by Noratrieb** [comment]
I see yeah, I left the comment a few minutes ago so it was with the latest state of your changes

**Comment by Noratrieb** [comment]
Thanks!
@bors r+

File: library/std/src/macros.rs [hunk 11]
```diff
@@ -379,3 +379,77 @@ macro_rules! dbg {
...
+#[allow_internal_unstable(hash_map_internals)]
+#[unstable(feature = "hash_map_macro", issue = "144032")]
+macro_rules! hash_map {
+    () => {{
+        $crate::collections::HashMap::new()
+    }};
+
+    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
+        let mut map = $crate::collections::HashMap::with_capacity(
+            const { $crate::repetition_utils!(@count $($key),*) }
```
Comment (danielhenrymantilla) [inline library/std/src/macros.rs:450]: I believe we should be able to simply do:

```suggestion
            ${count($key)}
```

File: library/std/src/macros.rs [hunk 12]
```diff
@@ -379,3 +379,77 @@ macro_rules! dbg {
...
+#[macro_export]
+#[allow_internal_unstable(hash_map_internals)]
+#[unstable(feature = "hash_map_internals", issue = "none")]
+macro_rules! repetition_utils {
+    (@count $($tokens:tt),*) => {{
+        [$($crate::repetition_utils!(@replace $tokens => ())),*].len()
+    }};
+
+    (@replace $x:tt => $y:tt) => { $y }
+}
```
Comment (danielhenrymantilla) [inline library/std/src/macros.rs:393]: In which case this would not be necessary:

```suggestion
```

File: library/std/src/macros.rs [hunk 13]
```diff
@@ -379,3 +379,77 @@ macro_rules! dbg {
...
+/// assert_eq!(map.get("key"), Some(&"value"));
+/// ```
+///
+/// The key and value are moved into the HashMap.
+///
+/// [`HashMap`]: crate::collections::HashMap
+/// [`HashMap::insert`]: crate::collections::HashMap::insert
+/// [`HashMap::with_capacity`]: crate::collections::HashMap::with_capacity
+#[macro_export]
+#[allow_internal_unstable(hash_map_internals)]
```
Comment (danielhenrymantilla) [inline library/std/src/macros.rs:441]: And you might then need:

```suggestion
#[allow_internal_unstable(hash_map_internals, macro_metavar_expr)]
```

File: library/std/src/macros.rs [hunk 14]
```diff
@@ -379,3 +379,77 @@ macro_rules! dbg {
...
+#[unstable(feature = "hash_map_macro", issue = "144032")]
+macro_rules! hash_map {
+    () => {{
+        $crate::collections::HashMap::new()
+    }};
+
+    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
```
Comment (danielhenrymantilla) [inline library/std/src/macros.rs:448]: Very tiny nit 🙂:

```suggestion
    ( $( $key:expr => $value:expr ),+ $(,)? ) => {{
```

would make the intent clearer, and also reject `hash_map!(,)`

File: library/std/src/macros.rs [hunk 15]
```diff
@@ -379,3 +379,77 @@ macro_rules! dbg {
...
+        let mut map = $crate::collections::HashMap::with_capacity(
+            const { $crate::repetition_utils!(@count $($key),*) }
+        );
+        $( map.insert($key, $value); )*
```
Comment (danielhenrymantilla) [inline library/std/src/macros.rs:452]:
```suggestion
        $( map.insert($key, $value); )+
```

**Review by danielhenrymantilla — COMMENTED** [review]
Looks good!

🤔 I wonder how the implementation would compare to doing:

```rs
<
    $crate::collections::HashMap<_, _, $crate::hash::RandomState>
    as
    $crate::convert::From<[(_, _); _]>
>::from([
    $(
        ($key, $value)
    ),+ // empty case still handled as `new()`.
])
```

Intuitively, doing a batch of `.insert()`s does require the `map` to go back to a fully usable-from-the-outside state, whereas a direct batched constructor such as `From<[(K, V); N]>` does not, so depending on the impl, there might be more room for optimisation with this more direct approach 🤔

File: library/std/src/macros.rs (hunk 11 above)
Comment (Noratrieb) [reply library/std/src/macros.rs:450]: you're a bit late :). the short answer is no, this is not easily possible without hacking up the feature gating code of that feature which is sort of annoying

File: library/std/src/macros.rs (hunk 13 above)
Comment (Noratrieb) [reply library/std/src/macros.rs:441]: doesn't work :)

**Comment by Noratrieb** [comment]
I have checked the `FromIterator` implementations, they are not any more efficient than repeat inserts (but the array impl will blow the stack :D)

**Comment by danielhenrymantilla** [comment]
Thanks for looking into these 🙏

File: library/std/src/macros.rs (hunk 14 above)
Comment (Noratrieb) [reply library/std/src/macros.rs:448]: that makes sense, I suggest addressing this in a follow-up PR

**Comment by stifskere** [comment]
> I have checked the `FromIterator` implementations, they are not any more efficient than repeat inserts (but the array impl will blow the stack :D)

I was now checking about this you said, which feels weird, I thought that without recursiveness this wouldn't have a problem, so should I add another PR to the tracking issue? Wait for a comment period?

**Comment by Noratrieb** [comment]
No what I mean is that if you made the macro expand to `from_iter([])` with all the values in an array then that array could get very large for large invocations and cause a stack overflow.
But with repeated insert, which you did, that's not a problem.

**Comment by stifskere** [comment]
Perfect. Thanks for everything @Noratrieb & @danielhenrymantilla !
